### PR TITLE
feat(low-code concurrent): [ISSUE #10550] have streams without partition routers nor cursor run …

### PR DESCRIPTION
**Pull request by maxi297**
…concurrently

## Summary by CodeRabbit

- **New Features**
	- Enhanced handling of concurrent streams with improved cursor management.
	- Simplified cursor creation logic for datetime incremental streams.
	- Added validation for primary key handling in stream configurations.

- **Bug Fixes**
	- Adjustments made to ensure accurate state tracking for concurrent and synchronous streams.
	- Improved error handling for cursor creation with invalid parameters.

- **Tests**
	- Updated test suite to validate new cursor logic and ensure robustness in handling various stream configurations, including expanded scenarios for concurrent state handling.
	- New tests added for primary key handling to cover valid and invalid input cases.

### CODE REVIEW

File: unit_tests/sources/declarative/test_concurrent_declarative_source.py [hunk 1]
```diff
@@ -479,20 +479,19 @@ def test_group_streams():
     synchronous_streams = source._synchronous_streams
 
     # 2 incremental streams
```
Comment (pnilan) [inline unit_tests/sources/declarative/test_concurrent_declarative_source.py:481]: nit: should update these comments to reflect 2 incremental, 1 full refresh as concurrent and 1 substream as synchronous

**Review by pnilan — APPROVED** [review]
Looks good!

Still need to test with an manifest-only connector.

File: airbyte_cdk/sources/streams/concurrent/helpers.py [hunk 2]
```diff
@@ -13,8 +13,13 @@ def get_primary_key_from_stream(
     elif isinstance(stream_primary_key, str):
         return [stream_primary_key]
     elif isinstance(stream_primary_key, list):
-        if len(stream_primary_key) > 0 and all(isinstance(k, str) for k in stream_primary_key):
+        are_all_elements_str = all(isinstance(k, str) for k in stream_primary_key)
+        are_all_elements_list_of_size_one = all(isinstance(k, list) and len(k) == 1 for k in stream_primary_key)
```
Comment (pnilan) [inline airbyte_cdk/sources/streams/concurrent/helpers.py]: Should we also verify that each list also contains a string?

`are_all_elements_list_of_size_one_and_str = all(isinstance(k, list) and len(k) == 1 and all(isinstance(j, str) for j in k) for k in test)`

File: airbyte_cdk/sources/streams/concurrent/helpers.py [hunk 3]
```diff
@@ -13,8 +13,13 @@ def get_primary_key_from_stream(
...
+
+        if are_all_elements_str:
             return stream_primary_key  # type: ignore # We verified all items in the list are strings
+        elif are_all_elements_list_of_size_one:
+            return list(map(lambda x: x[0], stream_primary_key))
```
Comment (pnilan) [inline airbyte_cdk/sources/streams/concurrent/helpers.py:24]: Does mypy fail here (like line 20) due to not knowing each list element contains a str?

**Review by pnilan — APPROVED** [review]
Two minor comments and then good to go.

File: airbyte_cdk/sources/streams/concurrent/helpers.py (hunk 2 above)
Comment (pnilan) [reply airbyte_cdk/sources/streams/concurrent/helpers.py]: > covered by the typing of stream_primary_key: Optional[Union[str, List[str], List[List[str]]]]

**Review by brianjlai — APPROVED** [review]
re-reviewing the new parts. the new retriever check and changes seems sane to me. :shipit:

**Review by darynaishchenko — APPROVED** [review]

**Pull request merged**